### ISSUE TITLE
Support dead letter queues

### ISSUE THREAD

**Issue by tcz** [php-amqplib/RabbitMqBundle#52]
RabbitMQ 2.8 implements AMQP's dead letter queue. 
http://lists.rabbitmq.com/pipermail/rabbitmq-announce/2012-March/000044.html

If would be nice to give support for this in the RabbitMqBundle.

**Comment by iLukey** [comment]
Hey all, 

I'm aware this is a very old issue, but the solution doesn't seem to cover all use-cases. 

In the scenario where my route is to publish a message to a fanout exchange, wherein one of a number of consumers fails, I want to dead-letter that message to the DLX - which is of type 'topic' - to a queue of a similar name as the original. 

The config format here seems to be restricted to just one queue per DLX. Moreover, the example given, the error_consumer is at the same level of nesting as the 'consumers' key, but it needs to fall under that level of nesting. 

Furthermore, there should ideally be more than one error_consumer block allowed as well, so that logical grouping can occur (allowing for multiple DLXs). This example might help explain what I mean:

I have two ways to create an order - the UI and an API. I have two inbound exchanges which are pushed to when an order has been created via either method: 

* order.created
* order.api.created

Both of these exchanges are of type fanout, and both have the same three queues bound to them:
* send_email
* send_sms
* add_note

In every scenario, let's assume that the latter two consumers work just fine, but sending an email is broken. After my TTL expires for each of them, they automatically get pushed to their respective DLX, order.created.dead_lettered or order.api.created.dead_lettered. 

Both of these exchanges are of type topic, and the messages are dead-lettered with the routing key 'send_email' - the same as the original queue. 

In each DLX, there are the following queues, and associated routing keys:
* send_email.failed - send_email
* send_sms.failed - send_sms
* add_note.failed - add_note

So, to summarize, the config would need to allow for:
* Multiple DLXs
* Multiple queues per DLX (to support type of 'topic' and 'fanout')

The reason for two exchanges is to logically separate the two types of message - despite them currently being routed to the same three queues. In future, they may warrant their own queues, or queues which are specific to the order creation source (UI or API) alongside the more generic queues (think if the message format is different between the two). 

As it stands, we have to specify one - and only one - DLX and DLX queue per config as I understand it, meaning that the other DLXs, and any additional DLX queues, need to be defined manually and wont be captured in the 'setup-fabric' call.

**Comment by Gyvastis** [comment]
> It supports dead lettering by configuration:
> 
> ```yaml
> old_sound_rabbit_mq:
>     producers:
>         work_producer:
>             connection: default
>             exchange_options: {name: work-exchange, type: direct}
> 
>     consumers:
>         work_consumer:
>             connection: default
>             exchange_options: {name: work-exchange, type: direct}
>             queue_options:    {name: work-queue, arguments: {'x-dead-letter-exchange': ['S', 'error-exchange']}}
>             callback: work_consumer_callback
> 
>     error_consumer:
>         connection: default
>         exchange_options: {name: error-exchange, type: direct}
>         queue_options:    {name: error-queue}
>         callback: error_consumer_callback
> ```
> 
> So if you publish a message with the work consumer and the message gets rejected, RabbitMQ will transfer the message to the error_queue.
> 
> Note: Currently you can not reject the message with requeue=false using this bundle. You can configure a TTL for the message but your consumer will redo the message until TTL is reached. An example for the above with a TTL of 5 sec.
> 
> ```yaml
> queue_options:    {name: work-queue, arguments: {'x-dead-letter-exchange': ['S', 'error-exchange'], 'x-message-ttl': ['I', 5000]}}
> ```
> 
> An improvement to reject messages with requeue=false is in this pull request https://github.com/videlalvaro/RabbitMqBundle/pull/109

Thanks for the comment. Spent half a day trying to figure out how to make it work with this bundle realising that it doesn't really work as per RabbitMQ documentation.
